File: solutions/3451-string-compression-iii/string-compression-iii.rb
def compressed_string(word)
    # Handle empty string case - return empty string if input is empty
    return "" if word.empty?
    
    # Initialize result string to store compressed output
    comp = ""
    
    # Initialize counter to track consecutive character occurrences
    cnt = 1
    
    # Initialize first character to process
    ch = word[0]
    
    # Iterate through string starting from second character
    (1...word.length).each do |i|
        # If current character matches previous and count less than 9
        # increment counter
        if word[i] == ch && cnt < 9
            cnt += 1
        # If character changes or count reaches 9
        # append count and character to result
        else
            comp << cnt.to_s << ch    # Convert count to string and append with character
            ch = word[i]              # Update current character being tracked
            cnt = 1                   # Reset counter for new character
        end
    end
    
    # Handle the last group of characters
    comp << cnt.to_s << ch    # Append final count and character
    
    # Return compressed string
    comp
end

File: solutions/3451-string-compression-iii/string-compression-iii.swift
class Solution {
    func compressedString(_ word: String) -> String {
        // Handle empty string case
        guard !word.isEmpty else { return "" }
        
        // Initialize result string
        var comp = ""
        comp.reserveCapacity(word.count * 2)
        
        // Initialize counter and first character
        var cnt = 1
        var ch = word.first!
        
        // Iterate through string starting from second character
        let chars = Array(word)
        for i in 1..<chars.count {
            // If current character matches previous and count < 9
            // increment counter
            if chars[i] == ch && cnt < 9 {
                cnt += 1
            } else {
                // If character changes or count reaches 9
                // append count and character to result
                comp.append(String(cnt))    // Convert count to string and append
                comp.append(ch)             // Append the character
                ch = chars[i]               // Update current character
                cnt = 1                     // Reset counter for new character
            }
        }
        
        // Handle the last group of characters
        comp.append(String(cnt))    // Append final count
        comp.append(ch)             // Append final character
        
        return comp
    }
}